doc/developers/contributing.rst

```diff
@@ -251,117 +242,133 @@ then submit a "pull request" (PR):
+   `non-regression tests
+   <https://en.wikipedia.org/wiki/Non-regression_testing>`_. These tests
+   verify the correct behavior of the fix or feature. In this manner, further
+   modifications on the code base are granted to be consistent with the
+   desired behavior. In the case of bug fixes, at the time of the PR, the
+   non-regression tests should fail for the code base in the master branch
+   and pass for the PR code.
+
+4. **Make sure that your PR does not add PEP8 violations**. On a Unix-like
+   system, you can run `make flake8-diff`. `flake8 path_to_file`, would work
+   for any system, but please avoid reformatting parts of the file that your
+   pull request doesn't change, as it distracts from code review.
+
+5. Follow the `coding-guidelines`_ (see below).
+
+6. When applicable, use the validation tools and scripts in the
+   ``sklearn.utils`` submodule.  A list of utility routines available
+   for developers can be found in the :ref:`developers-utils` page.
+
+7. Give your pull request a helpful title that summarises what your
+   contribution does. This title will often become the commit message once
+   merged so it should summarise your contribution for posterity. In some
+   cases "Fix <ISSUE TITLE>" is enough. "Fix #<ISSUE NUMBER>" is never a
+   good title.
+
+8. Often pull requests resolve one or more other issues (or pull requests).
+   If merging your pull request means that some other issues/PRs should
+   be closed, you should `use keywords to create link to them
+   <https://github.com/blog/1506-closing-issues-via-pull-requests/>`_
+   (e.g., ``Fixes #1234``; multiple issues/PRs are allowed as long as each
+   one is preceded by a keyword). Upon merging, those issues/PRs will
+   automatically be closed by GitHub. If your pull request is simply
+   related to some other issues/PRs, create a link to them without using
+   the keywords (e.g., ``See also #1234``).
+
+9. PRs should often substantiate the change, through benchmarks of
+   performance and efficiency or through examples of usage. Examples also
+   illustrate the features and intricacies of the library to users. Have a
+   look at other examples in the `examples/
+   <https://github.com/scikit-learn/scikit-learn/tree/master/examples>`_
+   directory for reference. Examples should demonstrate why the new
+   functionality is useful in practice and, if possible, compare it to other
+   methods available in scikit-learn.
+
+10. New features often need to be illustrated with narrative documentation in
+    the user guide, with small code snipets. If relevant, please also add
+    references in the literature, with PDF links when possible.
+
+11. The documentation should also include expected time and space complexity
+    of the algorithm and scalability, e.g. "this algorithm can scale to a
+    large number of samples > 100000, but does not scale in dimensionality:
+    n_features is expected to be lower than 100".
 
```

Comment:
It might also be instructive to point the contributor to the Code Review Guidelines to give some idea of what reviewers will be looking for.